lib/Sema/TypeOfReference.cpp

```diff
@@ -2487,18 +2487,19 @@ void ConstraintSystem::resolveOverload(ConstraintLocator *locator,
       }
     }
 
-    if (isa<SubscriptDecl>(decl)) {
-      if (locator->isResultOfKeyPathDynamicMemberLookup() ||
-          locator->isKeyPathSubscriptComponent()) {
-        // Subscript type has a format of (Self[.Type) -> (Arg...) -> Result
-        auto declTy = adjustedOpenedType->castTo<FunctionType>();
-        auto subscriptTy = declTy->getResult()->castTo<FunctionType>();
-        // If we have subscript, each of the arguments has to conform to
-        // Hashable, because it would be used as a component inside key path.
-        for (auto index : indices(subscriptTy->getParams())) {
-          const auto &param = subscriptTy->getParams()[index];
-          verifyThatArgumentIsHashable(index, param.getParameterType(), locator);
-        }
+    if ((isa<SubscriptDecl>(decl) &&
+         (locator->isResultOfKeyPathDynamicMemberLookup() ||
+          locator->isKeyPathSubscriptComponent())) ||
+        (isa<FuncDecl>(decl) && locator->isKeyPathMemberComponent())) {
+      // Subscript/method type has a format of (Self[.Type) -> (Arg...) ->
+      // Result
+      auto declTy = adjustedOpenedType->castTo<FunctionType>();
+      auto ty = declTy->getResult()->castTo<FunctionType>();
+      // If we have subscript or method, each of the arguments has to conform to
+      // Hashable, because it would be used as a component inside key path.
```

Comment:
Hm, I know that most of this is a preexisting code but I feel like we have a chance to improve the situation now. I think that the terms are mixed here - the code doesn't check arguments, it checks the parameters of the overload choice that it's about to be bound, but I think this check is actually about argument values because they are captured and would be applied at runtime, this means that we new `Apply` component we can introduce conformance constraints while synthesizing conforms for that component...